### DISCUSSION TITLE
Making ui.plot_card() visible and invisible

### DISCUSSION THREAD

**Discussion by n-srinidhi** [h2oai/wave#1830] [Q&A] [answered]
1. Is there a `visible` property for `ui.plot_card` to make it visible or invisible based on switching between menus or tabs?
2. Is it possible to make an entire zone visible or invisible based on events?
I am trying to switch between tabs and I need to hide and unhide certain components based on tab selection. But this would also mean that the layout of the displayed items will be changing too. So If I can organize them into different zones, and hide and unhide the zones it will be super useful.

**Comment by marek-mihok** [comment]
1. One option could be controlling the card visibility by its width. e.g.:
```py
box=ui.box('zone1', width=('100%' if visible else '0px'))
```
2. Could you please provide us with a specific use case of your app why isn't hiding and unhiding cards/components inside the existing zone enough?

**Reply by n-srinidhi** [reply]
Thank you! 
1. My current layout is like this:
```
ui.zone(name='header'),
ui.zone(name='main'),
ui.zone(name='content', direction='row',size='40%'),
ui.zone(name='c_tables',direction='row',size='40%'),
```
My `ui.tabs` is in the `main` zone, 2 `plot_card` instances in `content` zone and 3 `ui.table` instances in `c_tables`.
2. So I have two tabs, and on switching tabs I need to display different components. Like Tab 1 will have following layout:
![image](https://user-images.githubusercontent.com/117831190/218448827-92c05266-fb91-4bba-9147-e1361201ecd9.png)
Where the bottom row will have 3 tables, and on switching to tab 2, I need to show only onw drop down on the right corner and two tables in the page.

Doesn't plot_card have a `visible` property I can toggle as True and False through code as and when required?

**Reply by n-srinidhi** [reply]
@marek-mihok Executing this code `box=ui.box('zone1', width=('100%' if visible else '0px'))` as it is gives me an error that visible isnt recognized. Do you have any examples for this? I tried these two options as well but neither of them worked:

`q.page['tab1_mstsym'].plot_card.box=ui.box(zone='content',order=1,width='0px')`
`q.page['tab1_mstsym'].plot_card.box.width='0px'`

**Comment by mturoci** [comment]
> Is there a visible property for ui.plot_card to make it visible or invisible based on switching between menus or tabs?

There is not, you can only delete the plot cards using `del q.page['your_plot_card']`.

> Is it possible to make an entire zone visible or invisible based on events?

For navigation in general, you should delete the cards (`del q.page['card']`) that are part of let's say page1 when navigating to page2 and recreate them when you go back. Any filled in state will be lost.

If this is what you need, I suggest making a new toy project, create a venv, install wave and run `wave init`. Then pick random examples which have the navigation already solved and see how that's implemented. This is the recommended way as of today.

On the other hand, hiding is useful if you switch between let's say 2 parts of the UI like 2 tabs or something and want to keep the their existing state. For these, there is a `visible` prop on form components since tabs are part of the form_card usually.

**Reply by n-srinidhi** [reply]
Hi! Thank you for your response. So I tried using the `ui.visualization()` which is a Form component and has the visible property available. Problem is it requires me to set `pack=True` which means I cannot change the rows or underlying data buffer and need to recreate the card from scratch every time. Problem with recreating the card every time is that the cards blink for a split second on being recreated manually and that isn't flattering from a user experience perspective. Using plot card allows me to only change the data without recreating it. But the issue of not being able to hide it persists, Since I would like to preserve the states while switching. 
The `visible` property in ui.tabs will hide the entire tab right? How will the other tab be visible for switching?

**Reply by mturoci** [reply, marked as answer]
>requires me to set pack=True which means I cannot change the rows or underlying data buffer

Although not ideal, there is a way of doing the update - see https://github.com/h2oai/wave/issues/441#issuecomment-759656209.

> The visible property in ui.tabs will hide the entire tab right? How will the other tab be visible for switching?

The `visible` prop just toggless the component visibility it's attached to. If you change `visible` on `ui.tabs`, it just hides `ui.tabs`.

See this [example](https://wave.h2o.ai/docs/examples/form-visibility/) to get a better idea.
